src/Oscoin/Storage/Block/Memory.hs

```diff
@@ -0,0 +1,58 @@
+module Oscoin.Storage.Block.Memory
+    ( newBlockStoreIO
+    ) where
+
+import           Oscoin.Prelude
+
+import           Oscoin.Crypto.Blockchain hiding (lookupTx)
+import           Oscoin.Crypto.Blockchain.Block (Block)
+import qualified Oscoin.Crypto.Hash as Crypto
+import qualified Oscoin.Storage.Block.Abstract as Abstract
+import qualified Oscoin.Time.Chrono as Chrono
+
+import qualified Data.List as List
+
+newBlockStoreIO
+    :: (MonadIO m, Crypto.Hashable c tx)
+    => Chrono.NewestFirst [] (Block c tx (Sealed c s))
+    -> IO (Abstract.BlockStore c tx s m)
+newBlockStoreIO blks = do
+    blocksVar <- newMVar $ Chrono.toNewestFirst blks
+    let getBlocks = liftIO $ readMVar blocksVar
+    let modifyBlocks f = liftIO $ modifyMVar_ blocksVar $ pure . f
+    pure $ mkBlockStore getBlocks modifyBlocks
+
+mkBlockStore
+    :: (Monad m, Crypto.Hashable c tx)
+    => m [Block c tx (Sealed c s)]
+    -> (([Block c tx (Sealed c s)] -> [Block c tx (Sealed c s)]) -> m ())
+    -> Abstract.BlockStore c tx s m
+mkBlockStore getBlocks modifyBlocks = (blockStoreReader, blockStoreWriter)
+  where
+    blockStoreReader = Abstract.BlockStoreReader
+        { getGenesisBlock =
+            List.last <$> getBlocks
+        , lookupBlock = \h ->
+            find (\b -> blockHash b == h) <$> getBlocks
+        , lookupTx = \txHash -> lookupTx txHash <$> getBlocks
+        , getBlocksByDepth = \d ->
+              Chrono.NewestFirst . take (fromIntegral d) <$> getBlocks
+        , getBlocksByParentHash = \h ->
+              Chrono.NewestFirst . takeWhile (\b -> blockHash b /= h) <$> getBlocks
+        , getTip = List.head <$> getBlocks
+        }
+    blockStoreWriter = Abstract.BlockStoreWriter
+        { insertBlock = \b -> modifyBlocks $ \bs -> b:bs
```

Comment:
I might be wrong, but `modifyMVar` is not strict neither in the function it applies nor in the content it wraps, which means `b:bs` will leak memory as will build lots of thunks.

We should probably put a bang here, like `\!bs -> b:bs` or contemplate on a bigger scale the use of packages like [strict-concurrency](http://hackage.haskell.org/package/strict-concurrency)

Reply:
Why is it actually using an `MVar`? Can it not use a simple `IORef`? `atomicModifyIORef` (without the tick) is lazy too, though.

Reply:
Good point. I’ll switch to IORef